upgrading/topics/keycloak/changes-20_0_0.adoc

```diff
@@ -27,4 +27,89 @@ See the H2 documentation http://www.h2database.com/html/commands.html#set_non_ke
+
+= Deprecated methods from data providers and models were removed
+
+Prior to Keycloak 15, there was a clean-up of provider and model interfaces
+where we deprecated some methods. Javadoc of these methods contained
+a corresponding replacement method (see Javadoc of Keycloak 19 release).
+In this release the methods were removed. List of all changed classes is below.
+
+Most common patterns for deprecating and now removing the methods are
+the following.
+
+* Streamification - interfaces now contain only Stream-based methods.
++
+For example in https://github.com/keycloak/keycloak/blob/19.0.3/server-spi/src/main/java/org/keycloak/models/GroupProvider.java#L47-L57[`GroupProvider`] interface
++
+[source,java]
+----
+@Deprecated
+List<GroupModel> getGroups(RealmModel realm);
+----
+was replaced by
++
+[source,java]
+----
+Stream<GroupModel> getGroupsStream(RealmModel realm);
+----
++
+More details on streamification work can be found in
+https://issues.redhat.com/browse/KEYCLOAK-14011[KEYCLOAK-14011].
+
+* Consistent parameter ordering - methods now have strict parameter
+ordering where `RealmModel` is always the first parameter.
++
+For example in https://github.com/keycloak/keycloak/blob/19.0.3/server-spi/src/main/java/org/keycloak/storage/user/UserLookupProvider.java#L48-L52[`UserLookupProvider`] interface
++
+[source,java]
+----
+@Deprecated
+UserModel getUserById(String id, RealmModel realm);
+----
+was replaced by
++
+[source,java]
+----
+UserModel getUserById(RealmModel realm, String id)
+----
+
+== List of changed interfaces
+
+(`ok.` stands for `org.keycloak.` package)
+
+* `server-spi` module
+** `ok.credential.CredentialInputUpdater`
+** `ok.credential.UserCredentialStore`
+** `ok.models.ClientProvider`
+** `ok.models.ClientSessionContext`
+** `ok.models.GroupModel`
+** `ok.models.GroupProvider`
+** `ok.models.KeyManager`
+** `ok.models.KeycloakSessionFactory`
+** `ok.models.ProtocolMapperContainerModel`
+** `ok.models.RealmModel`
+** `ok.models.RealmProvider`
+** `ok.models.RoleContainerModel`
+** `ok.models.RoleMapperModel`
+** `ok.models.RoleModel`
+** `ok.models.RoleProvider`
+** `ok.models.ScopeContainerModel`
+** `ok.models.UserCredentialManager`
+** `ok.models.UserModel`
+** `ok.models.UserProvider`
+** `ok.models.UserSessionProvider`
+** `ok.models.utils.RoleUtils`
+** `ok.sessions.AuthenticationSessionProvider`
+** `ok.storage.client.ClientLookupProvider`
+** `ok.storage.group.GroupLookupProvider`
+** `ok.storage.user.UserLookupProvider`
+** `ok.storage.user.UserQueryProvider`
+* `server-spi-private` module
+** `ok.events.EventQuery`
+** `ok.events.admin.AdminEventQuery`
+** `ok.keys.KeyProvider`
+
+All changes can be found in the following https://github.com/keycloak/keycloak/pull/14973[pull request].
```

Comment:
I have not seen a PR in documentation before. Is this valid?

Reply:
Adding it there makes sense to me as it contains all the information about what was changed. On the other hand, if it is not common to add it to the upgrading guide I can remove it. @hmlnarik WDYT?

Reply:
Good catch. I agree with @andymunro here, it is common to refer to an issue, should it be needed. Could you please refer to https://github.com/keycloak/keycloak/issues/14720 instead? Something like:
```
All changes are linked to the following https://github.com/keycloak/keycloak/pull/14720[issue].
```